README.md
# Super Simple Call Blocker

## What is this?
Super Simple Call Blocker is a small executable written in JavaScript that aims to block all spam or unwanted calls directed to your landline phone.

It works by connecting to your ISP servers using the [SIP protocol](https://en.wikipedia.org/wiki/Session_Initiation_Protocol) and emulating a VOIP phone.

## How do I use it?
1. First, [download the latest release](https://github.com/josemmo/super-simple-call-blocker/releases/latest) from GitHub (Windows, Linux or Mac)
2. Create a valid configuration file (see next section for more info)
3. Run `sscb --config path-to-your-config-file.json` from a terminal

## How do I make it work?
A valid configuration file written in JSON is needed for the executable to be ran. Here is a basic example with all required fields needed to connect to a SIP server:

```javascript
{
  "server": "10.31.255.134",
  "port": 5070,
  "realm": "telefonica.net",
  "user": "911223344",
  "password": "911223344"
}
```

The rest optional fields with their respective default values are the following:

```javascript
{
  "expiration": 300, // Duration of SIP session in seconds. May be replaced by server.
  "expirationAdjust": -10, // Effective expiration timeout
  "maxForwards": 70, // Maximum jumps between routers (see SIP protocol)
  "debug": false, // Show or hide debug messages
  "userAgent": "SuperSimpleSIP/[version]"
}
```

## What numbers does it block?
At the moment, Super Simple Call Blocker connects to the following websites to check whether the phone is known to be a spam caller:

- [ListaSpam.com](https://www.listaspam.com)
- [esnumber.com](http://www.esnumber.com/)

It is also possible to specify a local blocklist of numbers by creating a text file:

```
# This is a text blocklist file
# Lines starting with pound sign are omitted
123456789
+34987654321
```

Note that the path to the blocklist needs to be set with `--blocklist path-to-the-blocklist.txt`.

## How do I keep track of my calls?
SSCB can keep a record of all incoming calls in plain text format. This option can be set by using `--log path-to-logfile.txt`.

## How do I compile it myself?
1. Download the source code using `npm install https://github.com/josemmo/super-simple-call-blocker.git`
2. Run `npm run build`


CHANGELOG.md
# Super Simple Call Blocker changelog

## v0.0.2
- Removed parameter `pingInterval` from settings as it is no longer necessary
- Added parameter `expirationAdjust` to settings
- Added parameter `maxForwards` to settings
- Fixed connected message loop bug
- Fixed SIP session loss bug

## v0.0.1
- Initial release


build.js
// Create output directory
const OUTPUT_DIR = __dirname + '/dist';
const fs = require('fs');
if (!fs.existsSync(OUTPUT_DIR)) fs.mkdirSync(OUTPUT_DIR);

// Define binaries to build
const version = require('./package.json').version;
const BINARIES = [
  {target: 'windows-x64-9.2.0', filename: `sscb-windows-x64-${version}.exe`},
  {target: 'windows-x86-9.2.0', filename: `sscb-windows-x86-${version}.exe`},
  {target: 'mac-x64-9.2.0', filename: `sscb-macos-x64-${version}`},
  {target: 'linux-x64-9.2.0', filename: `sscb-linux-x64-${version}`},
  {target: 'alpine-x86-9.2.0', filename: `sscb-alpine-x86-${version}`}
];

// Prepare to build
const nexe = require('nexe');
function build(i) {
  console.log(`[i] Building for ${BINARIES[i].target} . . .`);
  nexe.compile({
    input: './main.js',
    output: `${OUTPUT_DIR}/${BINARIES[i].filename}`,
    target: BINARIES[i].target,
    build: false
  }).then(() => {
    // Clear temporal files
    nexe.compile({
      clean: true,
      target: BINARIES[i].target
    }).then(() => {
      if ((i+1) < BINARIES.length) {
        build(i+1);
      } else {
        console.log('[i] Finished building sources');
        console.log('[i] Output directory: ' + OUTPUT_DIR);
      }
    });
  });
};

// Start building
console.log('[i] Generating binaries . . .');
build(0);


SuperSimpleSIP.js
function SuperSimpleSIP(CONFIG) {

  const STATUS = {"REGISTERING": 0, "IDLE": 1, "RINGING": 2, "HANGING_UP": 3};
  const ALLOW_METHODS = "PRACK, INVITE, ACK, BYE, CANCEL, UPDATE, INFO, " +
    "SUBSCRIBE, NOTIFY, REFER, MESSAGE, OPTIONS";
  const DEFAULTS = {
    expiration: 300, // Duration of SIP session in seconds. May be replaced by server.
    expirationAdjust: -10, // Effective expiration timeout
    maxForwards: 70, // Maximum jumps between routers (see SIP protocol)
    debug: false,
    userAgent: "SuperSimpleSIP/" + require('./package.json').version,
  };

  let socket = null, aliveTimeout = null, ts = this, lastSeqNum = 0;
  let caller = null, phoneStatus = STATUS.REGISTERING;
  let listeners = {
    connected: function() {},
    disconnected: function() {},
    error: function() {},
    incomingCall: function() {},
    callCanceled: function() {}
  };

  // Validate config
  for (let i in DEFAULTS) {
    if (typeof CONFIG[i] == 'undefined') CONFIG[i] = DEFAULTS[i];
  }


  /**
   * Get random string
   * @param  {int}    length Length
   * @param  {string} dict   Dictionary
   * @return {string}        String
   */
  function getRandomString(length, dict) {
    let res = "";
    for (let i=0; i<length; i++) {
      res += dict.charAt(Math.floor(Math.random() * dict.length));
    }
    return res;
  }


  /**
   * Get random branch
   * @return {string} Branch
   */
  function getRandomBranch() {
    return "z9hG4bK" + getRandomString(34, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
  }


  /**
   * Get random tag
   * @return {string} Tag
   */
  function getRandomTag() {
    return getRandomString(32, "0123456789abcdef");
  }


  /**
   * Log
   * @param  {string} message Message
   */
  function log(message) {
    if (CONFIG.debug) console.log(message);
  }


  /**
   * Send welcome message
   */
  function sendWelcomeMsg() {
    let addr = socket.address();
    let welcomeMsg = new Buffer(
      `REGISTER sip:${CONFIG.server}:${CONFIG.port} SIP/2.0\r\n` +
      `Via: SIP/2.0/UDP ${addr.address}:${addr.port};rport;branch=${getRandomBranch()}\r\n` +
      `Max-Forwards: ${CONFIG.maxForwards}\r\n` +
      `From: "${CONFIG.user}" <sip:${CONFIG.user}@${CONFIG.realm}>;tag=${getRandomTag()}\r\n` +
      `To: "${CONFIG.user}" <sip:${CONFIG.user}@${CONFIG.realm}>\r\n` +
      `Call-ID: ${getRandomTag()}\r\n` +
      `CSeq: ${lastSeqNum} REGISTER\r\n` +
      `User-Agent: ${CONFIG.userAgent}\r\n` +
      `Contact: "${CONFIG.user}" <sip:${CONFIG.user}@${addr.address}:${addr.port};ob>\r\n` +
      `Expires: ${CONFIG.expiration}\r\n` +
      `Allow: ${ALLOW_METHODS}\r\n` +
      'Content-Length: 0\r\n\r\n'
    );
    socket.send(welcomeMsg, CONFIG.port, CONFIG.server);
  }


  /**
   * Attach alive timeout
   */
  function attachAliveTimeout() {
    if (aliveTimeout !== null) clearTimeout(aliveTimeout);
    aliveTimeout = setTimeout(function() {
      restartConnection();
    }, (CONFIG.expiration+CONFIG.expirationAdjust)*1000);
  }


  /**
   * Restart connection
   */
  function restartConnection() {
    // Set variables to default
    if (phoneStatus != STATUS.REGISTERING) listeners.disconnected();
    caller = null;
    phoneStatus = STATUS.REGISTERING;
    lastSeqNum++;

    // Attach alive timeout
    attachAliveTimeout();


    // Wait for server to accept the connection
    let acceptInterval = setInterval(() => {
      if (phoneStatus == STATUS.REGISTERING) {
        sendWelcomeMsg();
      } else {
        clearInterval(acceptInterval);
      }
    }, 1500);

    // Send welcome message
    sendWelcomeMsg();
  }


  /**
   * Acknowledge invite
   */
  function ackInvite() {
    if (phoneStatus == STATUS.IDLE) {
      log('New caller: ' + JSON.stringify(caller));
      phoneStatus = STATUS.RINGING;
      let tryingMsg = new Buffer(
        'SIP/2.0 100 Trying\r\n' +
        `Via: ${caller.via}\r\n` +
        `Call-ID: ${caller.callID}\r\n` +
        `From: ${caller.from}\r\n` +
        `To: ${caller.to}\r\n` +
        `CSeq: ${caller.cSeq}\r\n` +
        'Content-Length: 0\r\n\r\n'
      );
      socket.send(tryingMsg, CONFIG.port, CONFIG.server, () => {
        ackInvite();
        listeners.incomingCall(caller.number);
      });
    } else if (phoneStatus == STATUS.RINGING) {
      let addr = socket.address();
      let ringingMsg = new Buffer(
        'SIP/2.0 180 Ringing\r\n' +
        `Via: ${caller.via}\r\n` +
        `Call-ID: ${caller.callID}\r\n` +
        `From: ${caller.from}\r\n` +
        `To: ${caller.to};tag=${caller.tag}\r\n` +
        `CSeq: ${caller.cSeq}\r\n` +
        `Contact: "${CONFIG.user}" <sip:${CONFIG.user}@${addr.address}:${addr.port};ob>\r\n` +
        `Allow: ${ALLOW_METHODS}\r\n` +
        'Content-Length: 0\r\n\r\n'
      );
      socket.send(ringingMsg, CONFIG.port, CONFIG.server);
    } else if (phoneStatus == STATUS.HANGING_UP) {
      ts.hangUp();
    }
  }


  /**
   * Connect
   */
  this.connect = function() {
    if (socket !== null) throw new Error('Phone already connected');
    lastSeqNum = Math.floor(Math.random() * 99999999);

    // Create UDP socket
    const dgram = require('dgram');
    socket = dgram.createSocket('udp4');

    // Be ready to listen for messages
    socket.on('message', (buffer, rinfo) => {
      let headers = buffer.toString().split('\r\n\r\n')[0];
      if (headers.indexOf('INVITE') === 0) {
        // Extract data from headers
        let via = headers.split('Via: ')[1].split('\r\n')[0];
        via = via.replace(CONFIG.port + ';', CONFIG.port + ';received=' +
          CONFIG.server + ';');
        let callID = headers.split('Call-ID: ')[1].split('\r\n')[0];
        let from = headers.split('From: ')[1].split('\r\n')[0];
        let to = headers.split('To: ')[1].split('\r\n')[0];
        let cSeq = headers.split('CSeq: ')[1].split('\r\n')[0];
        let number = headers.split('P-Asserted-Identity: <sip:')[1].split('@')[0];

        // Update caller information
        if ((caller === null) || (caller.number !== number)) {
          log('Caller number has changed');
          phoneStatus = STATUS.IDLE; // As it is a new caller
          caller = {number: number, tag: getRandomTag()};
        }
        caller.via = via;
        caller.callID = callID;
        caller.from = from;
        caller.to = to;
        caller.cSeq = cSeq;

        // Answer SIP server
        ackInvite();
      } else if (headers.indexOf('ACK') === 0) {
        log('Server answered with an acknowledgement');
        restartConnection();
      } else if (headers.indexOf('200 OK') > -1) {
        phoneStatus = STATUS.IDLE;
        let expiresArray = headers.match(/;expires=([0-9]+)/g);
        let lastExpiration = CONFIG.expiration;
        for (let i in expiresArray) {
          let timeout = Number(expiresArray[i].split('=')[1]);
          if (timeout < CONFIG.expiration) CONFIG.expiration = timeout;
        }
        if (lastExpiration != CONFIG.expiration) {
          log(`Changed session expiration from ${lastExpiration} to ` +
            `${CONFIG.expiration} seconds`);
        }
        attachAliveTimeout();
        listeners.connected();
      } else if (headers.indexOf('CANCEL') === 0) {
        if (caller !== null) {
          caller = null;
          listeners.callCanceled();
          log('Called canceled the call');
        }
        phoneStatus = STATUS.IDLE;
        let okSeq = headers.split('CSeq: ')[1].split('\r\n')[0];
        let okVia = headers.split('Via: ')[1].split('\r\n')[0];
        let okFrom = headers.split('From: ')[1].split('\r\n')[0];
        let okTo = headers.split('To: ')[1].split('\r\n')[0];
        let okCallID = headers.split('Call-ID: ')[1].split('\r\n')[0];
        let addr = socket.address();
        let okMsg = new Buffer(
          'SIP/2.0 200 OK\r\n' +
          `Via: ${okVia}\r\n` +
          `From: ${okFrom}\r\n` +
          `To: ${okTo}\r\n` +
          `Call-ID: ${okCallID}\r\n` +
          `CSeq: ${okSeq}\r\n` +
          'Content-Length: 0\r\n\r\n'
        );
        socket.send(okMsg, CONFIG.port, CONFIG.server);
      }
    });

    // Listen for errors
    socket.on('error', (err) => {
      listeners.error(err);
    });

    // Send welcome message
    socket.on('listening', () => {
      restartConnection();
    });

    // Open socket
    socket.bind();
  };


  /**
   * Hang up
   */
  this.hangUp = function() {
    if (caller === null) return;
    phoneStatus = STATUS.HANGING_UP;
    let addr = socket.address();
    let declineMsg = new Buffer(
      'SIP/2.0 603 Decline\r\n' +
      `Via: ${caller.via}\r\n` +
      `Call-ID: ${caller.callID}\r\n` +
      `From: ${caller.from}\r\n` +
      `To: ${caller.to};tag=${caller.tag}\r\n` +
      `CSeq: ${caller.cSeq}\r\n` +
      `Allow: ${ALLOW_METHODS}\r\n` +
      'Content-Length: 0\r\n\r\n'
    );
    socket.send(declineMsg, CONFIG.port, CONFIG.server);
  }


  /**
   * Disconnect
   */
  this.disconnect = function() {
    if (aliveTimeout !== null) {
      clearTimeout(aliveTimeout);
      aliveTimeout = null;
    }
    if (socket !== null) socket.close();
    caller = null;
    phoneStatus = STATUS.IDLE;
    listeners.disconnected();
  };


  /**
   * On
   * @param  {string}   evnt     Event
   * @param  {Function} callback Callback
   */
  this.on = function(evnt, callback) {
    if (typeof listeners[evnt] == 'undefined') throw new Error('Invalid event');
    if (typeof callback !== 'function') throw new Error('Not a function');
    listeners[evnt] = callback;
  };


  /**
   * Is debug
   * @return {boolean} Debug
   */
  this.isDebug = function() {
    return CONFIG.debug;
  }

}

module.exports = SuperSimpleSIP;


Blocklist/NumberUtils.js
let NumberUtils = (() => {

  const PREFIXES = [
    20, 27, 30, 31, 32, 33, 34, 36, 39, 40, 41, 43, 44, 45, 46, 47, 48, 49, 51,
    52, 53, 54, 55, 56, 57, 60, 61, 62, 63, 65, 66, 76, 77, 81, 86, 90, 91, 92,
    93, 94, 211, 212, 213, 216, 218, 220, 221, 222, 223, 224, 225, 226, 227,
    228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 240, 241, 244, 246,
    248, 249, 250, 251, 252, 253, 254, 256, 257, 258, 260, 261, 262, 263, 264,
    265, 266, 267, 268, 269, 291, 298, 299, 350, 351, 352, 353, 354, 355, 356,
    357, 358, 359, 370, 371, 372, 373, 374, 375, 376, 377, 378, 380, 381, 382,
    385, 386, 420, 421, 423, 501, 502, 503, 504, 505, 506, 507, 509, 590, 592,
    593, 594, 595, 596, 597, 598, 670, 673, 674, 675, 676, 677, 678, 679, 680,
    682, 683, 685, 686, 687, 688, 689, 690, 692, 852, 855, 880, 960, 961, 962,
    964, 965, 966, 967, 968, 971, 972, 973, 974, 975, 976, 977, 992, 993, 994,
    995, 996, 998, 1201, 1202, 1203, 1204, 1205, 1206, 1207, 1208, 1209, 1210,
    1212, 1213, 1214, 1215, 1216, 1217, 1218, 1219, 1224, 1225, 1226, 1228,
    1231, 1236, 1240, 1242, 1246, 1248, 1250, 1252, 1253, 1254, 1256, 1264,
    1267, 1270, 1281, 1289, 1301, 1302, 1303, 1304, 1305, 1306, 1307, 1308,
    1309, 1310, 1312, 1313, 1314, 1315, 1316, 1317, 1318, 1319, 1320, 1321,
    1323, 1330, 1334, 1336, 1345, 1347, 1352, 1360, 1361, 1401, 1402, 1403,
    1404, 1405, 1406, 1407, 1408, 1409, 1410, 1412, 1413, 1414, 1415, 1416,
    1417, 1418, 1419, 1423, 1424, 1425, 1431, 1435, 1438, 1440, 1441, 1443,
    1450, 1469, 1473, 1480, 1484, 1501, 1502, 1503, 1504, 1505, 1506, 1507,
    1508, 1509, 1510, 1512, 1513, 1514, 1515, 1516, 1517, 1518, 1519, 1520,
    1530, 1540, 1541, 1559, 1561, 1562, 1570, 1571, 1573, 1580, 1581, 1587,
    1601, 1602, 1603, 1604, 1605, 1606, 1607, 1608, 1609, 1610, 1612, 1613,
    1614, 1615, 1616, 1617, 1618, 1619, 1623, 1626, 1630, 1636, 1646, 1647,
    1650, 1651, 1660, 1661, 1662, 1664, 1669, 1670, 1671, 1678, 1684, 1701,
    1702, 1703, 1704, 1705, 1706, 1707, 1708, 1709, 1712, 1713, 1714, 1715,
    1716, 1717, 1718, 1719, 1720, 1724, 1727, 1732, 1734, 1740, 1757, 1758,
    1760, 1765, 1767, 1770, 1773, 1775, 1778, 1781, 1785, 1786, 1801, 1802,
    1803, 1804, 1805, 1806, 1807, 1810, 1812, 1813, 1814, 1815, 1816, 1817,
    1818, 1819, 1828, 1830, 1831, 1832, 1843, 1847, 1849, 1850, 1856, 1858,
    1860, 1864, 1867, 1870, 1876, 1901, 1902, 1903, 1904, 1905, 1906, 1907,
    1908, 1909, 1910, 1912, 1913, 1914, 1915, 1916, 1917, 1918, 1919, 1920,
    1925, 1931, 1935, 1937, 1939, 1940, 1941, 1949, 1952, 1954, 1956, 1970,
    1972, 1973, 1978, 6723, 262269, 262639, 441481, 441534, 6189162, 6189164
  ];


  /**
   * To local
   *
   * Given a phone number, converts it to a local phone number by removing
   * country prefix.
   *
   * @param  string number Phone number
   * @return string        Local phone number
   */
  function toLocal(number) {
    // Normalize prefix
    if (number.indexOf('00') === 0) number = '+' + number.substring(2);
    if (number.indexOf('+') !== 0) return number;

    // Detect prefix
    for (let i in PREFIXES) {
      if (number.indexOf('+' + PREFIXES[i]) === 0) {
        return number.substring(PREFIXES[i].toString().length + 1);
      }
    }

    // Fallback
    return number.substring(1);
  }


  // Return functions
  return {toLocal: toLocal};

})();

module.exports = NumberUtils;


Blocklist/FileBlocklist.js
function FileBlocklist(path) {

  let numbers = null;

  function readFile(callback) {
    const fs = require('fs');
    fs.readFile(path, 'utf8', function (err, data) {
      if (err) {
        console.error('Failed to open blocklist file');
        numbers = [];
      } else {
        numbers = data.replace(/\r/g, '').replace(/#(.+)\n/g, '').trim().split('\n');
      }
      callback();
    });
  }

  this.isSpam = function(number, callback) {
    if (numbers === null) {
      var ts = this;
      readFile(function() {
        ts.isSpam(number, callback);
      });
    } else {
      const numberParsed = require('./NumberUtils').toLocal(number);
      callback(
        (numbers.indexOf(number) > -1) || (numbers.indexOf(numberParsed) > -1)
      );
    }
  };

}

module.exports = FileBlocklist;


main.js
// Show welcome message
const version = require('./package.json').version;
console.log(
  "   _____                          _____ _                 _      \n" +
  "  / ____|                        / ____(_)               | |     \n" +
  " | (___  _   _ _ __   ___ _ __  | (___  _ _ __ ___  _ __ | | ___ \n" +
  "  \\___ \\| | | | '_ \\ / _ \\ '__|  \\___ \\| | '_ ` _ \\| '_ \\| |/ _ \\\n" +
  "  ____) | |_| | |_) |  __/ |     ____) | | | | | | | |_) | |  __/\n" +
  " |_____/ \\__,_| .__/ \\___|_|    |_____/|_|_| |_| |_| .__/|_|\\___|\n" +
  "        _____ | |  _ _   ____  _            _      | |           \n" +
  "       / ____||_| | | | |  _ \\| |          | |     |_|           \n" +
  "      | |     __ _| | | | |_) | | ___   ___| | _____ _ __        \n" +
  "      | |    / _` | | | |  _ <| |/ _ \\ / __| |/ / _ \\ '__|       \n" +
  "      | |___| (_| | | | | |_) | | (_) | (__|   <  __/ |          \n" +
  "       \\_____\\__,_|_|_| |____/|_|\\___/ \\___|_|\\_\\___|_|   " +
  `v${version}\n`);

// Read arguments
let config = null;
let localBlocklistPath = null;
let logPath = null;
process.argv.forEach((val, i) => {
  if (i == 0) return;
  if ((val == '--blocklist') || (val == '-b')) {
    localBlocklistPath = process.argv[i+1];
  } else if ((val == '--config') || (val == '-c')) {
    let data;
    try {
      data = require('fs').readFileSync(process.argv[i+1]);
    } catch (e) {
      data = process.argv[i+1];
    }
    config = JSON.parse(data);
  } else if ((val == '--log') || (val == '-l')) {
    logPath = process.argv[i+1];
  } else if ((val == '--help') || (val == '-h')) {
    console.log(
      '--help or -h: show this screen\n' +
      '--blocklist [filepath] or -b [filepath]: set blocklist local file\n\n' +
      '  NOTE: this file should be in plain text format, containing a\n' +
      '  phone number per line, as shown in this example:\n\n' +
      '  # This is a text blocklist file\n' +
      '  # Lines starting with pound sign are omitted\n' +
      '  123456789\n' +
      '  +34987654321\n\n' +
      '--config [json_string] or -c [json_string]: set configuration object\n' +
      '--config [filepath] or -c [filepath]: set configuration file\n\n' +
      '  NOTE: configuration file should be formatted as JSON and have,\n' +
      '  at least, the following fields as shown in this example:\n\n' +
      '  {\n' +
      '    "server": "10.31.255.134",\n' +
      '    "port": 5070,\n' +
      '    "realm": "telefonica.net",\n' +
      '    "user": "911223344",\n' +
      '    "password": "911223344"\n' +
      '  }\n\n' +
      '--log [filepath] or -l [filepath]: set log file for incoming calls\n\n'
    );
    process.exit(0);
  }
});
if (config === null) {
  console.error('Missing configuration. Run --help for more information.');
  process.exit(0);
}

// Define functions
function log(msg) {
  if (logPath === null) return;
  let date = (new Date()).toJSON();
  require('fs').appendFileSync(logPath, date + '\t' + msg + '\n');
}

// Create blocklists
const ListaSpamBlocklist = require('./Blocklist/ListaSpamBlocklist');
const EsnumberBlocklist = require('./Blocklist/EsnumberBlocklist');
const PoolBlocklist = require('./Blocklist/PoolBlocklist');
let blocklistArray = [new ListaSpamBlocklist(), new EsnumberBlocklist()];
if (localBlocklistPath !== null) {
  const FileBlocklist = require('./Blocklist/FileBlocklist');
  blocklistArray.unshift(new FileBlocklist(localBlocklistPath));
}
let blocklist = new PoolBlocklist(blocklistArray);

// Create new phone
const SuperSimpleSIP = require('./SuperSimpleSIP');
let phone = new SuperSimpleSIP(config);

// Detect errors
phone.on('error', () => {
  console.error('[!] Failed to connect to SIP server. Check configuration file.');
  process.exit(1);
});

// Listen for incoming calls
phone.on('incomingCall', (number) => {
  console.log(`[i] Incoming call from ${number} . . .`);
  blocklist.isSpam(number, function(isSpam) {
    console.log(`[i] Spam analysis result for ${number}: ${isSpam}`);
    log(number + '\t' + (isSpam ? 'SPAM' : 'HAM'));
    if (isSpam) phone.hangUp();
  });
});

// Detect disconnections
let isShuttingDown = false;
phone.on('disconnected', () => {
  if (!isShuttingDown && phone.isDebug()) {
    console.log('[i] Session expired. Trying to reconnect . . .');
  }
});

// Connect to SIP server
let isFirstConnection = true;
phone.on('connected', () => {
  if (isFirstConnection) {
    isFirstConnection = false;
    console.log('[i] Phone connected! Press "q" whenever you want to quit.');
  } else if (phone.isDebug()) {
    console.log('[i] Session renewed');
  }
});
console.log('[i] Now connecting . . .');
phone.connect();

// Listen to STDIN keypress
const stdin = process.stdin;
stdin.setRawMode(true);
stdin.resume();
stdin.setEncoding('utf8');
stdin.on('data', (key) => {
  if ((key === '\u0003') || (key === 'q') || (key === 'Q')) {
    console.log('[i] Now disconnecting . . .');
    isShuttingDown = true;
    phone.disconnect();
    process.exit(0);
  }
});
